### ISSUE TITLE
List group by AlbumArtistSort does not seem to fall back to AlbumArtist, Artist

### ISSUE THREAD

**Issue by geneticdrift** [MusicPlayerDaemon/MPD#355]
Not sure if this is the intended behavior but it's a bit confusing that when useing `list group` AlbumArtist falls back to Artist but AlbumArtistSort does not.

In my test db all files are tagged with Artist, only some are tagged with AlbumArtist, and none with AlbumArtistSort.

`list album group albumartist` falls back to Artist where there's no AlbumArtist but `list album group albumartistsort` just returns the list of Album tags.

MPD version

    Music Player Daemon 0.21 (v0.20.21-1399-gc5c4d4a57)

Tag types
```
tagtypes
tagtype: Artist
tagtype: ArtistSort
tagtype: Album
tagtype: AlbumSort
tagtype: AlbumArtist
tagtype: AlbumArtistSort
tagtype: Title
tagtype: Track
tagtype: Name
tagtype: Genre
tagtype: Date
tagtype: OriginalDate
tagtype: Composer
tagtype: Performer
tagtype: Disc
tagtype: MUSICBRAINZ_ARTISTID
tagtype: MUSICBRAINZ_ALBUMID
tagtype: MUSICBRAINZ_ALBUMARTISTID
tagtype: MUSICBRAINZ_TRACKID
tagtype: MUSICBRAINZ_RELEASETRACKID
tagtype: MUSICBRAINZ_WORKID
OK
```

**Comment by MaxKellermann** [comment]
I implemented the fallback, but uh-oh, `list group` is completely broken! It was broken from the start, but apparently nobody ever used it. Will be fixed in 0.20.22.

**Comment by MaxKellermann** [comment]
Broken `list group` fixed in commit db27bb76e280c69d70bc14f0b0161e89d496a3c8

**Comment by geneticdrift** [comment]
> apparently nobody ever used it

Just FYI, at least 3 android clients use the `list group` command.
It is the only way to disambiguate duplicate album names without doing a lot of work on the client side.

**Comment by geneticdrift** [comment]
I just tested a build from tag v0.20.22 and it looks like this fix has changed the meaning of group compared with previous versions.

The command `list album group albumartist` was returning album followed by albumartist so it was possible to differentiate albums with the same title but different album artist.
In v0.20.22 the same commands returns AlbumArtist first followed by albums.
It breaks client that are trying to parse according to the previous protocol behavior.

**Comment by MaxKellermann** [comment]
The old implementation didn't make any sense, and wasn't documented that way. It is unfortunate that this correction breaks clients depending on incorrect behavior, but these clients were buggy as well.

**Comment by geneticdrift** [comment]
What would be the correct way to get a list of unique albums?

The use case that `list album group albumartist` in its previous form answered is that of multiple albums with identical title but different album artist. It was a use case requested by users.

Also, I think MPD can benefit a lot from having SQLite as the backend for the music database.
It would make it much easier to implement command filters such as group, count, sort, and window in a consistent way for all the query-type commands of MPD. And also easier to maintain, and extend the protocol, and to present a more powerful, consistent, and clear commands for clients.

**Comment by geneticdrift** [comment]
Fair enough. I conflated a perceived problem with an implementation detail.
So putting SQLite aside the problems that I think make client's life difficult are that there's no consistency among all query commands, especially with regard to `window` `sort` `group` `count` modifiers.

Just as an example: the new corrected `list album group albumartist` to obtain unique albums is sorted by AlbumArtist not by album title.

**Comment by MaxKellermann** [comment]
The documentation doesn't specify that the result is sorted. Making assumptions on observed but undocumented behavior is going to get you into trouble eventually. Don't ever complain if your code breaks.
You may complain that the protocol documentation is too sparse and incomplete, and I'd agree. If you think something's missing in the documentation, then we should work towards completing it. But never rely on undocumented behavior.

**Comment by geneticdrift** [comment]
I should open a feature request for `sort` on arbitrary tags on all query commands and also `window` and `group`.
Ideally clients not have to pull some arbitrarily long list (sometimes hitting output buffer limit and kicked out) and then sort it. Instead get a sorted paged result they can fit to a scrolling list as a moving window (and they want the total count of the query too - with the same query args, and while at it they want to sum the duration field too without having to list all the tracks).
I guess that's too much to ask for. 
If I were tasked to implement all this I would certainly look for SQL first.
